### ISSUE TITLE
Properly maintain binary version of neard

### ISSUE THREAD

**Issue by bowenwang1996** [nearprotocol/nearcore#2677]
Currently we do not maintain the binary version of neard between releases and it has been staying at 0.4.13 for quite some time now. Since we are going with upgradability solution to avoid having hard forks, it is more important to properly maintain binary versions to avoid confusions. In terms of actual versioning, I suggest we follow https://semver.org/.

**Comment by chefsale** [comment]
I started working on a tool for managing releases. My proposal is that we don't cut a stable and beta branch as we do now, but just tag releases on the master. In the case of a hotfix we cut a new branch release the hotfix version and on merge on master release patches upstream to all major releases.

I like the  idea of using semver for the versioning format aka:

Given a version number MAJOR.MINOR.PATCH, increment the:

MAJOR version when you make incompatible API changes
**essentially a hard fork and similar**

MINOR version when you add functionality in a backwards compatible manner
** monthly stable releases **

PATCH version when you make backwards compatible bug fixes.
Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.
** security fixes and all the other patches in the meantime **

In this case we only cut a branch in a case of a PATCH which get merged into next minor release.

**Comment by mfornet** [comment]
> Examples: 1.0.0-devnet, 1, 1.0.0-betanet, 1.0.0-testnet, 1.0.0-othernet

I assume before phase 2 we will be using pre-releases `*-betanet`, `*-testnet`, ..., as releases practically.  In that case we should have a clear definition about when to promote one version from one network to the following more stable network.

**Comment by chefsale** [comment]
Yes, that is what I was thinking. As well even after phase 2, we can continue to run the same process backed by near foundation maybe?

**Comment by bowenwang1996** [comment]
As discussed with @chefsale, it is better to tag with `alpha` and `beta` since we are going to decouple nearcore releases from network update schedule. Some ideas:
* A version starts with `version-beta` in master (a release is cut)
* The version gets tested on betanet for some predefined amount of time.
* The version becomes a release candidate.
* The version gets tested on testnet for some predefined amount of time.
* The version is finalized.

**Comment by chefsale** [comment]
Essentially what I started working on is a tool which first of all tags a commit:

git log
... * - * -x -> y -> z -> master

release phases:
1.0.1-beta (nightly) -> 1.0.1-rc -> 1.0.1

Nightly runs at x tags the commit 1.0.1-beta (nightly) -> runs on betanet -> updates the tag and the release to 1.0.1-rc-> runs on testnet -> propagates the same commit to 1.0.1

In case of a failure we just push the new beta upstream and ignore the next propagation, aka if we figure out beta is broken we don't create a release candidate, etc.

**Comment by chefsale** [comment]
As well the flow would be:

We bump the version in https://github.com/nearprotocol/nearcore/blob/master/neard/Cargo.toml to generate manually a stable release using the pipeline I'm working on for Github Releases https://buildkite.com/nearprotocol/release-nearcore-manual which generated: https://github.com/nearprotocol/nearcore/releases/tag/1.0.0.

**Comment by bowenwang1996** [comment]
@chefsale sounds good. I believe the version bump and stable release generation can potentially be automated if we provide in advance a summary of the release. It can be done once the release is cut and since there is some time between when the release is cut and when it is finalized, it should be doable.

**Comment by chefsale** [comment]
Yes, in near-ops I have created a tool which should be able to do the automatic promotions and bumping. The folder is rear, it's still in progress. As well I'm working on creating automation of binary uploads to the github releases.
